Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp

```diff
@@ -0,0 +1,259 @@
+}
+
+void IOWindow::Update()
+{
+  m_expression_text->setPlainText(QString::fromStdString(m_reference->expression));
+  m_range_spinbox->setValue(m_reference->range * SLIDER_TICK_COUNT);
+  m_range_slider->setValue(m_reference->range * SLIDER_TICK_COUNT);
+
+  m_devq.FromString(m_controller->default_device.ToString());
+
+  UpdateDeviceList();
+  UpdateOptionList();
+}
+
+void IOWindow::ConnectWidgets()
+{
+  connect(m_select_button, &QPushButton::clicked, [this] { AppendSelectedOption(""); });
+  connect(m_add_button, &QPushButton::clicked, [this] { AppendSelectedOption(" + "); });
+  connect(m_and_button, &QPushButton::clicked, [this] { AppendSelectedOption(" & "); });
+  connect(m_or_button, &QPushButton::clicked, [this] { AppendSelectedOption(" | "); });
+  connect(m_not_button, &QPushButton::clicked, [this] { AppendSelectedOption("!"); });
+
+  connect(m_type == IOWindow::Type::Input ? m_detect_button : m_test_button, &QPushButton::clicked,
+          this, &IOWindow::OnDetectButtonPressed);
+
+  connect(m_button_box, &QDialogButtonBox::clicked, this, &IOWindow::OnDialogButtonPressed);
+  connect(m_devices_combo, &QComboBox::currentTextChanged, this, &IOWindow::OnDeviceChanged);
+  connect(m_range_spinbox, static_cast<void (QSpinBox::*)(int value)>(&QSpinBox::valueChanged),
+          this, &IOWindow::OnRangeChanged);
+  connect(m_range_slider, static_cast<void (QSlider::*)(int value)>(&QSlider::valueChanged), this,
+          &IOWindow::OnRangeChanged);
+}
+
+void IOWindow::AppendSelectedOption(const std::string& prefix)
+{
+  if (m_option_list->currentItem() == nullptr)
+    return;
+
+  m_expression_text->insertPlainText(
+      QString::fromStdString(prefix) +
+      MappingCommon::GetExpressionForControl(m_option_list->currentItem()->text(), m_devq,
+                                             m_controller->default_device));
+}
+
+void IOWindow::OnDeviceChanged(const QString& device)
+{
+  m_devq.FromString(device.toStdString());
+  UpdateOptionList();
+}
+
+void IOWindow::OnDialogButtonPressed(QAbstractButton* button)
+{
+  if (button == m_clear_button)
+  {
+    m_expression_text->clear();
+    return;
+  }
+
+  m_reference->expression = m_expression_text->toPlainText().toStdString();
+
+  if (button != m_apply_button)
+    accept();
+}
+
+void IOWindow::OnDetectButtonPressed()
+{
+  if (m_block)
+    return;
+
+  m_block = true;
+  std::thread([this] {
+    auto* btn = m_type == IOWindow::Type::Input ? m_detect_button : m_test_button;
+    const auto old_label = btn->text();
+
+    btn->setText(QStringLiteral("..."));
+
+    const auto expr = MappingCommon::DetectExpression(
+        m_reference, g_controller_interface.FindDevice(m_devq).get(), m_devq, m_devq);
+
+    btn->setText(old_label);
+
+    if (!expr.isEmpty())
+    {
+      const auto list = m_option_list->findItems(expr, Qt::MatchFixedString);
+
+      if (list.size() > 0)
+        m_option_list->setCurrentItem(list[0]);
+    }
+    m_block = false;
+  }).detach();
```

Comment:
It is correct, though not a good idea. This is how the existing `MappingButton` code was, though, so might be better to fix it all in one fell swoop in a separate PR.